packages/vite/src/node/preview.ts

```diff
@@ -48,12 +48,28 @@ export async function preview(
 
   const options = config.server
   const hostname = resolveHostname(serverOptions.host ?? options.host)
-  const port = serverOptions.port ?? 5000
+  let port = serverOptions.port ?? 5000
   const protocol = options.https ? 'https' : 'http'
   const logger = config.logger
   const base = config.base
 
+  const onError = (e: Error & { code?: string }) => {
+    if (e.code === 'EADDRINUSE') {
+      logger.info(`Port ${port} is in use, trying another one...`)
```

Comment:
Maybe we could count `serverOptions.port` set as strict and therefore have a similar logic like in dev mode.
So only when using the default port `5000` it will be incremented but an error will be thrown if set explicitly

Reply:
Yes, my current implementation is that if `serverOptions.port` is set, the `error` event will not be registered, so the port does not increase when it is used.
https://github.com/vitejs/vite/blob/5381db61b90f30f9c10f754e1127171bfcbcbbae/packages/vite/src/node/preview.ts#L51
https://github.com/vitejs/vite/blob/5381db61b90f30f9c10f754e1127171bfcbcbbae/packages/vite/src/node/preview.ts#L66-L68


I tried to extract a shared function, Is this ok？
```javascript
function httpServerStart(
    httpServer: HttpServer,
    serverOptions: { port: number; strict: boolean; host: string },
    fun: Function
  ) {
    let { port, strict, host } = serverOptions
    const onError = (e: Error & { code?: string }) => {
      if (strict) {
        httpServer.removeListener('error', onError)
        new Error(`Port ${port} is already in use`)
      } else if (e.code === 'EADDRINUSE') {
        logger.info(`Port ${port} is in use, trying another one...`)
        httpServer.listen(++port, host)
      } else {
        httpServer.removeListener('error', onError)
        throw e
      }
    }

    httpServer.listen(port, host, () => {
      httpServer.removeListener('error', onError)
      fun()
    })
  }
```

In `preview`:
```javascript
httpServerStart(
      httpServer,
      { port, strict: !!serverOptions.port, host: hostname.host },
      () => {
        //dosomething
      }
    )
```

In `dev server`
```javascript
httpServerStart(
      httpServer,
      { port, strict: options.strictPort, host: hostname.host },
      () => {
        //dosomething
      }
    )
```

Reply:
Looks good, just use `fun: () => void` instead

And if you like, you could use async-await instead of callback 🤔

Reply:
@Shinigami92 
Help review the code.